presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java

```diff
@@ -13,30 +13,64 @@
  */
 package com.facebook.presto.sql.planner.optimizations;
 
+import com.facebook.presto.sql.planner.iterative.Lookup;
 import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
 import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
 import com.facebook.presto.sql.planner.plan.ExchangeNode;
 import com.facebook.presto.sql.planner.plan.FilterNode;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 import com.facebook.presto.sql.planner.plan.PlanVisitor;
 import com.facebook.presto.sql.planner.plan.ProjectNode;
 import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.sql.tree.SymbolReference;
 import com.google.common.collect.ImmutableList;
 
+import java.util.function.Function;
+
 import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.function.Function.identity;
 
 public final class ScalarQueryUtil
 {
     private ScalarQueryUtil() {}
 
+    /**
+     * @return true when subquery is scalar and its output symbols are directly mapped to the ApplyNode's output symbols
+     */
+    public static boolean isResolvedScalarSubquery(ApplyNode applyNode)
+    {
+        return isScalar(applyNode.getSubquery(), x -> x) && applyNode.getSubqueryAssignments().getExpressions().stream()
```

Comment:
Other code using lookup functions use `identity()`. See `isScalar` in the same class